### PR TITLE
Feature/validate qnr

### PR DESCRIPTION
**Pull request by pbugni**
Significant changes to the swagger for QuestionnaireResponses, moved all the sub-schema objects up to the `defines` list.  Validation is now working, and we will no longer permit invalid QuestionnaireResponse submissions via PUT or POST.

### CODE REVIEW

File: portal/migrations/versions/69af4bd9ac9a_.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    """Attempt to fix the invalid structure of a QNR.document
+
+    Many current QNRs are invalid as a user submitted w/o the `answer` list
+    as requires in FHIR DSTU2 QuestionnaireResponse.
+
+    Commits fix and audit row to db
+
+    """
+    qnr = QuestionnaireResponse.query.get(qnr_id)
+    document = copy.deepcopy(qnr.document)
```
Comment (ivan-c) [inline portal/migrations/versions/69af4bd9ac9a_.py]: :+1:

File: portal/migrations/versions/69af4bd9ac9a_.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    answer_list = document['group']
+    if not answer_list:
+        raise ValueError("no answers under group for {}".format(qnr_id))
+
+    # each answer is to be a list of value_x_ objects
+    # nest in a list
+    for a in answer_list:
+        the_answer = a['answer']
+
+        # remove bogus 'system' at wrong level
```
Comment (ivan-c) [inline portal/migrations/versions/69af4bd9ac9a_.py]: I love atomicity of each of these individual fixes. Would it be worthwhile to create an audit entry for each "fix" ran to make it obvious what's changed to an outsider?

File: portal/migrations/versions/69af4bd9ac9a_.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+def upgrade():
+    """Iterate over all QNRs, validating and patching any that failed"""
+
+    ids_to_patch = []
+    validate = validate_closure()
+    sys.stderr.write("invalid QNR.ids: ")
+    for qnr in QuestionnaireResponse.query.all():
+        if not validate(qnr.document):
+            if 'question' in qnr.document['group']:
+                sys.stderr.write(
```
Comment (ivan-c) [inline portal/migrations/versions/69af4bd9ac9a_.py]: Why not use `logger`?

File: portal/views/assessment_engine.py [hunk 4]
```diff
@@ -1314,6 +1344,119 @@ def assessment_add(patient_id):
             questionnaire:
               display: EPIC 26 Short Form
               reference: https://stg.us.truenth.org/api/questionnaires/epic26
+      - schema:
```
Comment (ivan-c) [inline portal/views/assessment_engine.py]: Out of scope for this PR, but how would you feel about a static YAML (or JSON) file/directory to keep our schemas in (at least for the FHIR schemas)? Would make our view code more readable

**Comment by ivan-c** [comment]
@pbugni Going to merge `develop` to hopefully fix the tests

File: portal/migrations/versions/69af4bd9ac9a_.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+            ids_to_patch.append(qnr.id)
+            sys.stderr.write(" {}".format(qnr.id))
+        else:
+            sys.stdout.write('.')
+            sys.stdout.flush()
+
+    sys_user = User.query.filter_by(email='__system__').first()
+    sys_id = sys_user.id
+
+    for id in ids_to_patch:
```
Comment (ivan-c) [inline portal/migrations/versions/69af4bd9ac9a_.py]:
```suggestion
    for qnr_id in ids_to_patch:
```

File: portal/migrations/versions/69af4bd9ac9a_.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        correct_invalid_document(id, validate, sys_id)
```
Comment (ivan-c) [inline portal/migrations/versions/69af4bd9ac9a_.py]:
```suggestion
        correct_invalid_document(qnr_id, validate, sys_id)
```

File: portal/views/assessment_engine.py (hunk 4 above)
Comment (pbugni) [reply portal/views/assessment_engine.py]: not opposed - i'd like to clean up the code by removing these massive comment blocks.  small fear they won't be as well maintained, but that's the life of a comment.  validation will always help with that.

**Pull request merged**